internal/controller/pkg/signature/reconciler.go

```diff
@@ -0,0 +1,435 @@
+		return reconcile.Result{}, nil
+	}
+
+	ic, vc, err := r.config.ImageVerificationConfigFor(ctx, p.GetSource())
+	if err != nil {
+		log.Debug("Cannot get image verification config", "error", err)
+		p.SetConditions(v1.VerificationIncomplete(errors.Wrap(err, "cannot get image verification config")))
+		_ = r.client.Status().Update(ctx, p)
+		return reconcile.Result{}, errors.Wrap(err, "cannot get image verification config")
+	}
+	if vc == nil || vc.Cosign == nil {
+		// No verification config found for this image, so, we will skip
+		// verification.
+		log.Debug("No signature verification config found for image, skipping verification")
+		p.SetConditions(v1.VerificationSkipped())
+		return reconcile.Result{}, errors.Wrap(r.client.Status().Update(ctx, p), "cannot update package status")
+	}
+
+	ref, err := name.ParseReference(p.GetSource(), name.WithDefaultRegistry(r.registry))
+	if err != nil {
+		log.Debug("Cannot parse package image reference", "error", err)
+		p.SetConditions(v1.VerificationIncomplete(errors.Wrap(err, errParseReference)))
+		_ = r.client.Status().Update(ctx, p)
+		return reconcile.Result{}, errors.Wrap(err, errParseReference)
+	}
+
+	pullSecrets := make([]string, 0, 2)
+	for _, s := range p.GetPackagePullSecrets() {
+		pullSecrets = append(pullSecrets, s.Name)
+	}
+
+	_, s, err := r.config.PullSecretFor(ctx, p.GetSource())
+	if err != nil {
+		log.Debug("Cannot get image config pull secret for image", "error", err)
+		p.SetConditions(v1.VerificationIncomplete(errors.Wrap(err, "cannot get image config pull secret for image")))
+		_ = r.client.Status().Update(ctx, p)
+		return reconcile.Result{}, errors.Wrap(err, "cannot get image config pull secret for image")
+	}
+	if s != "" {
+		pullSecrets = append(pullSecrets, s)
+	}
+
+	auth, err := k8schain.New(ctx, r.clientset, k8schain.Options{
+		Namespace:          r.namespace,
+		ServiceAccountName: r.serviceAccount,
+		ImagePullSecrets:   pullSecrets,
+	})
+	if err != nil {
+		log.Debug("Cannot create k8s auth chain", "error", err)
+		p.SetConditions(v1.VerificationIncomplete(errors.Wrap(err, "cannot create k8s auth chain")))
+		_ = r.client.Status().Update(ctx, p)
+		return reconcile.Result{}, errors.Wrap(err, "cannot create k8s auth chain")
+	}
+
+	if err = r.validator.Validate(ctx, ref, vc, remote.WithAuthFromKeychain(auth)); err != nil {
+		log.Debug("Signature verification failed", "error", err)
+		p.SetConditions(v1.VerificationFailed(ic, []error{err}))
+		if err = r.client.Status().Update(ctx, p); err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "cannot update status with failed verification")
+		}
+		return reconcile.Result{}, errors.Wrap(err, "signature verification failed")
+	}
+
+	p.SetConditions(v1.VerificationSucceeded(ic))
+	return reconcile.Result{}, errors.Wrap(r.client.Status().Update(ctx, p), "cannot update status with successful verification")
+}
+
+func enqueueProvidersForImageConfig(kube client.Client, log logging.Logger) handler.EventHandler {
```

Comment:
These enqueue functions are all the same apart from the list type, right?

In https://github.com/crossplane/crossplane/pull/6019 I handled this by:

* Introducing a `ProviderList` interface, like `ProviderRevisionList`
* Accepting an instance of the type to list as an argument - `list v1.ProviderList`
* Making a deep copy of the provided instance and listing into that

Reply:
Great!

Can I just defer a refactor to there or do you suggest doing it here? I believe there are other places in the codebase already which needs to refactored, so, my preference would be doing them all together as a follow up.

Reply:
I don't think my PR will affect these ones specifically, but it would prevent you needing to do all the `ProviderList` work. I'm happy either way.

Reply:
Oh, actually, we have the PackageRevisionList interface already.

Just refactored them into a single function following your deep copy trick.